lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -16,30 +17,48 @@ module Cardano.Wallet.Primitive.AddressDiscovery.Random
     (
     -- ** State
       RndState (..)
-    , mkRndState
     ) where
 
 import Prelude
 
+import Cardano.Byron.Codec.Cbor
+    ( decodeAddressDerivationPath, decodeAddressPayload, deserialise )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..), XPrv )
 import Cardano.Wallet.Primitive.AddressDerivation.Random
-    ( RndKey )
+    ( RndKey (..) )
 import Cardano.Wallet.Primitive.AddressDiscovery
     ( CompareDiscovery (..)
     , GenChange (..)
     , IsOurs (..)
     , IsOwned (..)
     , KnownAddresses (..)
     )
-import Data.ByteString
-    ( ByteString )
+import Cardano.Wallet.Primitive.Types
+    ( Address (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeDeserialiseCbor, unsafeFromHex )
+import Control.DeepSeq
+    ( NFData )
+import GHC.Generics
+    ( Generic )
 
-newtype RndState = RndState ByteString
+import qualified Data.ByteString.Lazy as BL
 
-mkRndState :: RndState
-mkRndState = error "AddressDiscovery.Random unimplemented"
+
+newtype RndState = RndState { getRndState :: RndKey 'RootK XPrv }
+    deriving (Generic)
+
+instance NFData RndState
 
 instance IsOurs RndState where
-    isOurs _ s = (False, s)
+    isOurs (Address addr) (RndState key) = do
+        let payload = unsafeDeserialiseCbor decodeAddressPayload $
+                BL.fromStrict (unsafeFromHex addr)
```

Comment:
No! The `addr` in an `Address` is a raw bytestring that isn't encoded in any particular format!! 
I'd suggest the following:

```hs
isOurs (Address payload) (RndState key) = do
    let payload = deserialise decodeAddressPayload payload
```

Reply:
Fixed ... maybe it was the change to base16 encoding in the golden test addresses that confused things.

Reply:
Yeah. I actually did the change precisely because a first version of the PR included base58 encoding. In the end, it comes from a confusion between the address binary format and user-facing encoding. I've tried to clarify that distinction in the wiki.